cli.go

```diff
@@ -0,0 +1,502 @@
+	}
+
+	fmt.Fprintf(os.Stderr, "Complete the login via your OIDC provider. Launching browser to:\n\n    %s\n\n\n", authURL)
+	if err := openURL(authURL); err != nil {
+		fmt.Fprintf(os.Stderr, "Error attempting to automatically open browser: '%s'.\nPlease visit the authorization URL manually.", err)
+	}
+
+	// Set up callback handler
+	http.HandleFunc(fmt.Sprintf("/v1/auth/%s/oidc/callback", mount), func(w http.ResponseWriter, req *http.Request) {
+		var response string
+
+		query := req.URL.Query()
+		code := query.Get("code")
+		state := query.Get("state")
+		data := map[string][]string{
+			"code":  {code},
+			"state": {state},
+		}
+
+		secret, err := c.Logical().ReadWithData(fmt.Sprintf("auth/%s/oidc/callback", mount), data)
+		if err != nil {
+			summary, detail := parseError(err)
+			response = errorHTML(summary, detail)
+		} else {
+			response = successHTML
+		}
+
+		w.Write([]byte(response))
+		doneCh <- loginResp{secret, err}
+	})
+
+	// Start local server
+	go func() {
+		if err := http.ListenAndServe(":"+port, nil); err != nil && err != http.ErrServerClosed {
+			fmt.Fprintf(os.Stderr, "Error listening for callback: %v\n\n", err.Error())
+		}
+	}()
+
+	// Wait for either the callback to finish or SIGINT to be received
+	select {
+	case s := <-doneCh:
+		return s.secret, s.err
+	case <-ch:
+		return nil, errors.New("interrupted")
+	}
+}
+
+func fetchAuthURL(c *api.Client, role, mount, port string) (*api.Secret, error) {
+	data := map[string]interface{}{
+		"role":         role,
+		"redirect_uri": fmt.Sprintf("http://localhost:%s/v1/auth/%s/oidc/callback", port, mount),
+	}
+
+	return c.Logical().Write(fmt.Sprintf("auth/%s/oidc/auth_url", mount), data)
+}
+
+// openURL opens the specified URL in the default browser of the user.
+// Source: https://stackoverflow.com/a/39324149/453290
+func openURL(url string) error {
+	var cmd string
+	var args []string
+
+	switch runtime.GOOS {
+	case "windows":
+		cmd = "cmd"
+		args = []string{"/c", "start"}
+	case "darwin":
+		cmd = "open"
+	default: // "linux", "freebsd", "openbsd", "netbsd"
+		cmd = "xdg-open"
+	}
+	args = append(args, url)
+	return exec.Command(cmd, args...).Start()
+}
+
+// parseError converts error from the API into summary and detailed portions.
+func parseError(err error) (string, string) {
+	headers := []string{errNoResponse, errLoginFailed, errTokenVerification}
+	summary := "Login error"
+	detail := ""
+
+	re := regexp.MustCompile(`(?s)Errors:.*\* *(.*)`)
+
+	errorParts := re.FindStringSubmatch(err.Error())
+	switch len(errorParts) {
+	case 0:
+		summary = ""
+	case 1:
+		detail = errorParts[0]
+	case 2:
+		for _, h := range headers {
+			if strings.HasPrefix(errorParts[1], h) {
+				summary = h
+				detail = strings.TrimSpace(errorParts[1][len(h):])
+				break
+			}
+		}
+		if detail == "" {
+			detail = errorParts[1]
+		}
+	}
+
+	return summary, detail
+
+}
+
+func errorHTML(summary, detail string) string {
```

Comment:
Since these are so long and not Go, would it be possible to drop them into separate files? It would be super intuitive to name them something like `error.html.tmpl` and to consider using [Go's native template parsing](https://golang.org/pkg/text/template/).

Reply:
Bare template files don't work too well here since they'd have to be loaded at runtime but aren't bundled with the binary. So you're back to Go files for the strings. But I do agree that they're not like the rest of the CLI Go code and have moved them into their own file.

Reply:
Sounds good! Thank you!